### PR TITLE
Added skeleton for PSA backend.

### PR DESCRIPTION
**Pull request by hanw**
init implementation for generating header/structs.

### CODE REVIEW

File: backends/bmv2/bmv2.cpp [hunk 1]
```diff
@@ -84,12 +84,26 @@ int main(int argc, char *const argv[]) {
         return 1;
 
     // backend depends on the modified refMap and typeMap from midEnd.
-    BMV2::Backend backend(options.isv1(), &midEnd.refMap,
-            &midEnd.typeMap, &midEnd.enumMap);
+    BMV2::Backend backend(options, &midEnd.refMap, &midEnd.typeMap, &midEnd.enumMap);
+
+    if (auto arch = getenv("P4C_DEFAULT_ARCH")) {
```
Comment (mbudiu-vmw) [inline backends/bmv2/bmv2.cpp]: This kind of pattern suggests to me that you actually have a hidden subtype hidden: you probably want a base class for backend with two subclasses for SS and PSA.

File: backends/bmv2/action.cpp [hunk 2]
```diff
@@ -18,128 +18,6 @@ limitations under the License.
 
 namespace BMV2 {
 
-void
-ConvertActions::convertActionBody(const IR::Vector<IR::StatOrDecl>* body, Util::JsonArray* result) {
```
Comment (mbudiu-vmw) [inline backends/bmv2/action.cpp]: Where has all this code moved?

File: backends/bmv2/CMakeLists.txt [hunk 3]
```diff
@@ -74,16 +72,16 @@ add_library(bmv2backend ${BMV2_BACKEND_SRCS})
 add_dependencies(bmv2backend genIR)
 
 build_unified(BMV2_SRCS)
-add_executable(p4c-bm2-ss ${BMV2_SRCS})
-target_link_libraries (p4c-bm2-ss bmv2backend ${P4C_LIBRARIES} ${P4C_LIB_DEPS})
+add_executable(p4c-bmv2 ${BMV2_SRCS})
```
Comment (mbudiu-vmw) [inline backends/bmv2/CMakeLists.txt]: Can we keep around a symbolic link or a wrapper script to the old program?
We don't want to break people's toolchains unnecessarily.

File: backends/bmv2/portableSwitch.h [hunk 4]
```diff
@@ -30,76 +30,152 @@ limitations under the License.
...
+    void postorder(const IR::P4Parser *p) override;
+    void postorder(const IR::P4Control* c) override;
+    void postorder(const IR::Type_Header* h) override;
+    void postorder(const IR::Type_HeaderUnion* hu) override;
+    void postorder(const IR::Declaration_Variable* var) override;
+    void postorder(const IR::Declaration_Instance* di) override;
+    void postorder(const IR::P4Action* act) override;
+    void postorder(const IR::Type_Error* err) override;
+
+    // control
```
Comment (mbudiu-vmw) [inline backends/bmv2/portableSwitch.h]: these comments are not particularly useful. I think you could remove them.

**Comment by hanw** [comment]
I am sorry, I opened this PR by mistake, I was working with another student on a new PSA backend for bmv2. The PR was intended to go against my fork. We were thinking about making a PR once we are done. But if it makes review easier, we can submit smaller PRs as we move long.

**Pull request closed without merging**